This is a reorganizing of selecting core after the discussion with JC and Lester August 2021. The output datasets are using the date in 20210818 and later updated as 20210825.

In [2]:
import numpy as np 
import pandas as pd
import glob
import matplotlib.pyplot as plt

#plt.style.use('ggplot')
#plt.style.use('seaborn-colorblind')
plt.style.use('dark_background')
plt.rcParams['figure.dpi'] = 300
plt.rcParams['savefig.dpi'] = 300
plt.rcParams['savefig.bbox'] = 'tight'

import datetime
date = datetime.datetime.now().strftime('%Y%m%d')

# Check data
if we have all the data on the list (ML station list.xlsx).

In [2]:
xl_df = pd.read_excel('data/ML station list.xlsx', sheet_name='CHOSEN')
xl_df

Station    latitude     longitude water depth (m)
0   LV29-114-3    49.375667  152.877933             1762
1   SO178-12-3    51.605367     145.4357            1211
2   SO264-13-2     37.79775    170.72211            3935
3   SO264-15-2     41.85472    170.51111            3662
4   SO264-28-2     45.00166    170.32305            1935
5   SO264-55-1     47.32027     169.4975            2936
6   SO264-56-2     47.94222    169.77722            3973
7   SO264-64-1     50.12055    168.34666            3495
8   SO264-66-2     50.26861    168.51138            2751
9   PS97-027-2   -54.384833   -74.605833         2341.8 
10  PS97-046-4   -60.997167   -65.356167         2775.3 
11  PS97-052-4      -62.499   -64.293667         2890.4 
12  PS97-053-2   -62.662833     -63.0935         2016.1 
13  PS97-078-1     -60.6525   -55.839667         3666.4 
14  PS97-079-2   -60.142333   -58.991333         3541.3 
15  PS97-080-1   -59.674667   -59.631333         3105.9 
16  PS97-083-2   -58.994333   -60.570333         3762.3 
17  PS97-084-1      -58.869   -60.865667           3557 
18  PS97-085-3     -58.3545   -62.167167         3090.8 
19  PS97-089-1   -58.226667     -62.7265         3437.2 
20  PS97-092-1      -57.763   -69.878833         3823.7 
21  PS97-093-2   -57.499167   -70.274667         3782.2 
22  PS97-128-2  -53.634330     -75.54583         2313.4 
23  PS75-054-1    -56.15175  -115.133033            4113
24  PS75-083-1   -60.268833  -159.059833            3599
25  PS75-093-1   -60.872167  -169.548167            3762
26  PS75-095-5   -57.019667      -174.43            4853

In [20]:
import glob

for core in xl_df.Station:
    print('Core {} has: {}'.format(
        core, 
        glob.glob('data/Bulk chem/{}*'.format(core))))

Core LV29-114-3 has: ['data/Bulk chem/LV29-114-3_TOC%_CaCO3%LMax.xls']
Core SO178-12-3 has: ['data/Bulk chem/SO178-12-3_TOC.txt']
Core SO264-13-2 has: ['data/Bulk chem/SO264-13-2_bulk.xlsx']
Core SO264-15-2 has: ['data/Bulk chem/SO264-15-2_bulk.xlsx']
Core SO264-28-2 has: ['data/Bulk chem/SO264-28-2_bulk.xlsx']
Core SO264-55-1 has: ['data/Bulk chem/SO264-55-1_bulk.xlsx']
Core SO264-56-2 has: ['data/Bulk chem/SO264-56-2_bulk.xlsx']
Core SO264-64-1 has: ['data/Bulk chem/SO264-64-1_bulk.xlsx']
Core SO264-66-2 has: ['data/Bulk chem/SO264-66-2_bulk.xlsx']
Core PS97-027-2 has: ['data/Bulk chem/PS97-027-2_bulk CaCO3.xlsx']
Core PS97-046-4 has: ['data/Bulk chem/PS97-046-4_bulk CaCO3.xlsx']
Core PS97-052-4 has: ['data/Bulk chem/PS97-052-4_bulk CaCO3.xlsx']
Core PS97-053-2 has: ['data/Bulk chem/PS97-053-2_bulk CaCO3.xlsx']
Core PS97-078-1 has: ['data/Bulk chem/PS97-078-1_bulk CaCO3.xlsx']
Core PS97-079-2 has: ['data/Bulk chem/PS97-079-2_bulk CaCO3.xlsx']
Core PS97-080-1 has: ['data/Bulk chem/PS9

# Check previous data

In [15]:
pd.read_csv('data/spe+bulk_PS75-056-1_20210601.csv', index_col=0)

0    1    2    3    4    5    6    7    8    9  ...  2041  2042  2043  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   ...   
98   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
99   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
100  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
101  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
102  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   

     2044  2045  2046  2047  CaCO3%        core  mid_depth_mm  
0     0.0   0.0   0.0   0.0  78.870  PS75-056-1          10.0  
1     0.0   0.0   0.0   0.0  77.889  PS75-056-1          50.0  
2     0.0   0.0   0.0   0.0  75.332  PS75-056-1         150.0  
3     0.0   0.0   0.0   0.0  76.673  PS75-056-1         250.0  
4     0.0   0.0   0.0   0.0  73.165  PS75-056-1         350.0  
..    ...   ...   ...   ...     ...         ...           ...  
98    0.0   0.0   0.0   0.0  24.091  PS75-056-1        9750.0  
99    0.0   0.0   0.0   0.0  26.568  PS75-056-1        9850.0  
100   0.0   0.0   0.0   0.0  28.045  PS75-056-1        9950.0  
101   0.0   0.0   0.0   0.0  24.707  PS75-056-1       10050.0  
102   0.0   0.0   0.0   0.0  23.696  PS75-056-1       10150.0  

[103 rows x 2051 columns]

In [16]:
pd.read_csv('data/spe+bulk_SO264-15-2.csv', index_col=0)

0    1    2    3    4    5    6    7    8    9  ...  2043  2044  2045  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
2   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
5   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
6   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
7   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
8   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
9   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
10  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
11  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
13  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
14  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
15  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
16  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
17  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
18  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
19  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
20  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
23  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
24  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
25  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
26  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
27  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
28  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
29  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
30  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
31  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
32  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
33  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
34  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
35  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
36  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
37  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
38  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
39  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   

    2046  2047       cps  mid_depth_cm  toc_percent  tc_percent  caco3_percent  
0    0.0   0.0  153268.5          10.5     0.181999    8.000254      65.149516  
1    0.0   0.0  122791.5          20.5     0.298746    6.007120      47.567882  
2    0.0   0.0  133795.0          46.5     0.288175    6.747430      53.824967  
3    0.0   0.0  152620.5          68.5     0.299631    6.199101      49.160281  
4    0.0   0.0  160492.0          88.5     0.240458    7.642422      61.680569  
5    0.0   0.0  158137.5         160.5     0.186681    9.317410      76.086367  
6    0.0   0.0  108823.5         190.5     0.377708    4.152185      31.452720  
7    0.0   0.0  141913.5         240.5     0.198560    8.380529      68.180347  
8    0.0   0.0   95248.0         250.5     0.234415    2.788586      21.283909  
9    0.0   0.0  137865.5         290.5     0.125437    7.886049      64.669186  
10   0.0   0.0  113905.0         

In [17]:
pd.read_csv('data/spe+bulk_dataset_20201215.csv', index_col=0)

0    1    2    3    4    5    6    7    8    9  ...  2043  2044  2045  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   ...   
735  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
737  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
738  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
739  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
740  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   

     2046  2047       TC%      TOC%     CaCO3%        core  mid_depth_mm  
0     0.0   0.0  2.542079  0.394127  17.898887  SO264-64-1         115.0  
1     0.0   0.0  2.247150  0.611208  13.632300  SO264-64-1         215.0  
2     0.0   0.0  0.710588  0.523402   1.559822  SO264-64-1         305.0  
3     0.0   0.0  0.562171  0.472551   0.746802  SO264-64-1        1015.0  
4     0.0   0.0  0.578167  0.312852   2.210866  SO264-64-1        1815.0  
..    ...   ...       ...       ...        ...         ...           ...  
735   0.0   0.0       NaN  0.710000        NaN  SO178-12-3        5715.0  
737   0.0   0.0       NaN  0.730000        NaN  SO178-12-3        5915.0  
738   0.0   0.0       NaN  0.710000        NaN  SO178-12-3        6015.0  
739   0.0   0.0       NaN  0.730000        NaN  SO178-12-3        6215.0  
740   0.0   0.0       NaN  0.760000        NaN  SO178-12-3        6415.0  

[699 rows x 2053 columns]

In [3]:
pre_df = pd.read_csv('data/spe+bulk_dataset_20201215.csv', index_col=0)
print(pre_df[pre_df.core == 'SO264-15-2'].head())
print(pre_df[pre_df.core == 'SO264-15-2'].shape)

       0    1    2    3    4    5    6    7    8    9  ...  2043  2044  2045  \
309  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
310  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
311  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
312  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
313  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   

     2046  2047       TC%      TOC%     CaCO3%        core  mid_depth_mm  
309   0.0   0.0  8.000254  0.181999  65.149516  SO264-15-2         105.0  
310   0.0   0.0  6.007120  0.298746  47.567882  SO264-15-2         205.0  
311   0.0   0.0  6.747430  0.288175  53.824967  SO264-15-2         465.0  
312   0.0   0.0  6.199101  0.299631  49.160281  SO264-15-2         685.0  
313   0.0   0.0  7.642422  0.240458  61.680569  SO264-15-2         885.0  

[5 rows x 2053 columns]
(38, 2053)


I can't find records why the core SO264-15-2 has its spe and spe+bulk dataset separately, but it's also included into the dataset_20201215, which used for building models for the manuscript at the early half of 2021. I will just use the dataset_20201215 as the "previous."

In [4]:
pre_df[pre_df.core == 'PS75-056-1']

Empty DataFrame
Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]
Index: []

[0 rows x 2053 columns]

The core PS75-056-1 need to be merged with dataset_20201215 as the previous datasets.

In [4]:
print(len(pre_df))
pre_df = pd.concat([pre_df, pd.read_csv('data/spe+bulk_PS75-056-1_20210601.csv', index_col=0)],
                   join='outer', axis=0)
print(len(pre_df))

699
802


In [5]:
pre_df.head()

0    1    2    3    4    5    6    7    8    9  ...  2043  2044  2045  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...   0.0   0.0   0.0   

   2046  2047       TC%      TOC%     CaCO3%        core  mid_depth_mm  
0   0.0   0.0  2.542079  0.394127  17.898887  SO264-64-1         115.0  
1   0.0   0.0  2.247150  0.611208  13.632300  SO264-64-1         215.0  
2   0.0   0.0  0.710588  0.523402   1.559822  SO264-64-1         305.0  
3   0.0   0.0  0.562171  0.472551   0.746802  SO264-64-1        1015.0  
4   0.0   0.0  0.578167  0.312852   2.210866  SO264-64-1        1815.0  

[5 rows x 2053 columns]

# Check and build datasets for the new cores

In [6]:
xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]

9     PS97-027-2
10    PS97-046-4
11    PS97-052-4
12    PS97-053-2
13    PS97-078-1
14    PS97-079-2
15    PS97-080-1
16    PS97-083-2
17    PS97-084-1
18    PS97-085-3
19    PS97-089-1
20    PS97-092-1
21    PS97-093-2
22    PS97-128-2
23    PS75-054-1
24    PS75-083-1
25    PS75-093-1
26    PS75-095-5
Name: Station, dtype: object

These are the cores listed in the excel but not yet into our dataset.

## Build spectral dataset

### Common cores

In [7]:
file_name = []
spe_all = []
depth_all = []
cps_all = []
core_all = []
s_depth_all = []

# open an error txt for recording errors
with open('results/error_log.txt', 'w') as ef:
    # do it core by core
    for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
        
        # only read the 10kV which having better signal to the light elements
        spe_dir = glob.glob('data/{}/Run 1 at  10kV/*.spe'.format(core))
        
        # make sure the order follows the depthes in filename
        spe_dir.sort()
        print('{} has {} data points.'.format(core, len(spe_dir)))

        for spe in spe_dir:
                   
            check_depth = spe.split()[3].split('_')[-1]
            
            # there are some inconsistencies in nameing...as usual
            # 5 and 6 digis means in mm
            try:
                if len(check_depth) >= 5:
                    start_depth = int(check_depth)
                # 3 and 4 digit means in cm, needs to be multipled to be mm
                elif len(check_depth) >= 3:
                    start_depth = int(check_depth) * 10

                file_name.append(spe.split('/')[-1])
                
                with open(spe, 'r') as f:
                    content = []
                    lines = f.readlines()
                    for line in lines[49:]:
                        content = np.hstack((content, line.split()))
                    
                    section_depth = int(lines[13][:-3])
                    spe_all.append(content.astype(int))
                    cps_all.append(int(lines[28]))
                    core_all.append(core)
                    s_depth_all.append(section_depth)
                    depth_all.append(section_depth + start_depth)
            except ValueError :
                print('Error in filename: {}'.format(spe), file=ef)

        print('core {} is done.'.format(core))

PS97-027-2 has 198 data points.
core PS97-027-2 is done.
PS97-046-4 has 126 data points.
core PS97-046-4 is done.
PS97-052-4 has 298 data points.
core PS97-052-4 is done.
PS97-053-2 has 187 data points.
core PS97-053-2 is done.
PS97-078-1 has 210 data points.
core PS97-078-1 is done.
PS97-079-2 has 789 data points.
core PS97-079-2 is done.
PS97-080-1 has 105 data points.
core PS97-080-1 is done.
PS97-083-2 has 695 data points.
core PS97-083-2 is done.
PS97-084-1 has 1067 data points.
core PS97-084-1 is done.
PS97-085-3 has 1443 data points.
core PS97-085-3 is done.
PS97-089-1 has 1000 data points.
core PS97-089-1 is done.
PS97-092-1 has 789 data points.
core PS97-092-1 is done.
PS97-093-2 has 3237 data points.
core PS97-093-2 is done.
PS97-128-2 has 163 data points.
core PS97-128-2 is done.
PS75-054-1 has 2232 data points.
core PS75-054-1 is done.
PS75-083-1 has 1308 data points.
core PS75-083-1 is done.
PS75-093-1 has 1271 data points.
core PS75-093-1 is done.
PS75-095-5 has 1783 data

No other error apart from the ValueError, which is recorded in the error_log.txt and skiped, pops up. After checking error_log.txt, I found only core PS97-083-2 has inconsistent naming spectra. The spectrum filename has TC in the position I adopt for section depth, e.g. PS97-083-2_"00TC"   20.0mm  10s  10kV 150uA No-Filter   0.0mm Run1 Rep0 Sett Low.spe. They should be identified as "Triger Core", but since there are no bulk chemistry for this TC core, I just abandon these data.

Luckily, these cores have the same spe format as the previous cores, which confirms they were scanned by the new scanning head.

### Triger cores

In [8]:
# no initialization because I want the data to be appended
# open an error txt for appending errors
with open('results/error_log.txt', 'a') as ef:
    # do it core by core
    # these are the cores having TC as a subfolder
    for core in ['PS97-084-1', 'PS97-085-3', 'PS97-089-1', 'PS75-093-1', 'PS75-083-1']:
        
        # only read the 10kV which having better signal to the light elements
        spe_dir = glob.glob('data/{}/TC/Run 1 at  10kV/*.spe'.format(core))
        
        # make sure the order follows the depthes in filename
        spe_dir.sort()
        print('{} has {} data points.'.format(core, len(spe_dir)))

        for spe in spe_dir:
                   
            check_depth = spe.split()[3].split('_')[-1]
            
            # there are some inconsistencies in nameing...as usual
            # 5 and 6 digis means in mm
            try:
                if len(check_depth) >= 5:
                    start_depth = int(check_depth)
                # 3 and 4 digit means in cm, needs to be multipled to be mm
                elif len(check_depth) >= 3:
                    start_depth = int(check_depth) * 10
            
            except ValueError:
                # There is only PS75-093-1 has TC in the naming and their sectiondepth should
                # all be starting at 0
                start_depth = 0

            file_name.append(spe.split('/')[-1])
            
            with open(spe, 'r') as f:
                content = []
                lines = f.readlines()
                for line in lines[49:]:
                    content = np.hstack((content, line.split()))
                
                section_depth = int(lines[13][:-3])
                spe_all.append(content.astype(int))
                cps_all.append(int(lines[28]))
                core_all.append(core+'_TC')   # mark as the triger cores
                s_depth_all.append(section_depth)
                depth_all.append(section_depth + start_depth)
            
        print('core {} is done.'.format(core))

PS97-084-1 has 89 data points.
core PS97-084-1 is done.
PS97-085-3 has 89 data points.
core PS97-085-3 is done.
PS97-089-1 has 90 data points.
core PS97-089-1 is done.
PS75-093-1 has 92 data points.
core PS75-093-1 is done.
PS75-083-1 has 37 data points.
core PS75-083-1 is done.


In [9]:
spe_df = pd.DataFrame(spe_all, columns = [str(_) for _ in range(2048)])
spe_df['cps'] = cps_all
spe_df['core'] = core_all
spe_df['composite_depth_mm'] = depth_all
spe_df['section_depth_mm'] = s_depth_all
spe_df['filename'] = file_name

print(spe_df.head())
print(spe_df.shape)

   0  1  2  3  4  5  6  7  8  9  ...  2043  2044  2045  2046  2047     cps  \
0  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0     0   61661   
1  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0     0  113413   
2  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0     0  116326   
3  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0     0  109808   
4  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0     0  109130   

         core  composite_depth_mm  section_depth_mm  \
0  PS97-027-2                  40                40   
1  PS97-027-2                  50                50   
2  PS97-027-2                  60                60   
3  PS97-027-2                  70                70   
4  PS97-027-2                  80                80   

                                            filename  
0  PS97-027-2_0000   40.0mm  10s  10kV 150uA No-F...  
1  PS97-027-2_0000   50.0mm  10s  10kV 150uA No-F...  
2  PS97-027-2_0000   60.0mm  10s  1

In [10]:
spe_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17208 entries, 0 to 17207
Columns: 2053 entries, 0 to filename
dtypes: int64(2051), object(2)
memory usage: 269.5+ MB


In [11]:
spe_df[spe_df.isna().any(axis=1)]

Empty DataFrame
Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]
Index: []

[0 rows x 2053 columns]

### Build composite_id

In [12]:
spe_df.composite_depth_mm.max()

22320

In [13]:
composite_id = []
for core, depth in zip(spe_df.core, spe_df.composite_depth_mm):
    composite_id.append('{}_{:05}'.format(core, depth))
    
spe_df['composite_id'] = composite_id

### Drop duplicates

In [15]:
clean_df = spe_df.drop_duplicates('composite_id', keep = 'last')
len(clean_df)

17187

### Check those duplicates

In [16]:
spe_df.loc[spe_df.composite_id.duplicated(keep = 'last'), spe_df.columns[-6:]]

cps        core  composite_depth_mm  section_depth_mm  \
1307  110515  PS97-079-2                4290               880   
1308  110283  PS97-079-2                4300               890   
1309  107649  PS97-079-2                4310               900   
1310  107895  PS97-079-2                4320               910   
1311  108733  PS97-079-2                4330               920   
1312  108215  PS97-079-2                4340               930   
1313  107611  PS97-079-2                4350               940   
1314  107120  PS97-079-2                4360               950   
1315  104004  PS97-079-2                4370               960   
1316  102841  PS97-079-2                4380               970   
1317  101885  PS97-079-2                4390               980   
1318  102622  PS97-079-2                4400               990   
1319  103659  PS97-079-2                4410              1000   
1320   68447  PS97-079-2                4420              1010   
3190  111536  PS97-084-1                7950               950   
3191  104955  PS97-084-1                7960               960   
3192  110855  PS97-084-1                7970               970   
3193  111098  PS97-084-1                7980               980   
3194  110839  PS97-084-1                7990               990   
3195  105549  PS97-084-1                8000              1000   
5827   56605  PS97-089-1                8020              1010   

                                               filename      composite_id  
1307  PS97-79-2_0341  880.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04290  
1308  PS97-79-2_0341  890.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04300  
1309  PS97-79-2_0341  900.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04310  
1310  PS97-79-2_0341  910.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04320  
1311  PS97-79-2_0341  920.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04330  
1312  PS97-79-2_0341  930.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04340  
1313  PS97-79-2_0341  940.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04350  
1314  PS97-79-2_0341  950.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04360  
1315  PS97-79-2_0341  960.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04370  
1316  PS97-79-2_0341  970.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04380  
1317  PS97-79-2_0341  980.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04390  
1318  PS97-79-2_0341  990.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04400  
1319  PS97-79-2_0341 1000.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04410  
1320  PS97-79-2_0341 1010.0mm  10s  10kV 150uA No-Fi...  PS97-079-2_04420  
3190  PS97-084-1_0700  950.0mm  10s  10kV 150uA No-F...  PS97-084-1_07950  
3191  PS97-084-1_0700  960.0mm  10s  10kV 150uA No-F...  PS97-084-1_07960  
3192  PS97-084-1_0700  970.0mm  10s  10kV 150uA No-F...  PS97-084-1_07970  
3193  PS97-084-1_0700  980.0mm  10s  10kV 150uA No-F...  PS97-084-1_07980  
3194  PS97-084-1_0700  990.0mm  10s  10kV 150uA No-F...  PS97-084-1_07990  
3195  PS97-084-1_0700 1000.0mm  10s  10kV 150uA No-F...  PS97-084-1_08000  
5827  PS97-089-1_0701 1010.0mm  10s  10kV 150uA No-F...  PS97-089-1_08020

Just some overlaps at section edges. I would simply delete them.

In [17]:
clean_df = clean_df.set_index('composite_id')

### Build section no.

In [19]:
section_all = []
# make sure the order follows the core and composite depth
clean_df.sort_values(by = 'composite_id', axis = 0, inplace = True)

for core in np.unique(clean_df.core):    
    # I assume every core scanned from section 0 so the first section in the core 
    # is marked as section 0. The deeper the larger number
    section = 0
    X = clean_df.loc[clean_df.core == core, 'section_depth_mm']
    for i in range(len(X)):
        section_all.append(section)
        try:
            # when section changes, the section depth should be reset to smaller number
            if X[i] > X[i + 1]:
                section += 1
        except IndexError:
            print('bottom of the core {}'.format(core))
        
clean_df['section'] = section_all

bottom of the core PS75-054-1
bottom of the core PS75-083-1
bottom of the core PS75-083-1_TC
bottom of the core PS75-093-1
bottom of the core PS75-093-1_TC
bottom of the core PS75-095-5
bottom of the core PS97-027-2
bottom of the core PS97-046-4
bottom of the core PS97-052-4
bottom of the core PS97-053-2
bottom of the core PS97-078-1
bottom of the core PS97-079-2
bottom of the core PS97-080-1
bottom of the core PS97-083-2
bottom of the core PS97-084-1
bottom of the core PS97-084-1_TC
bottom of the core PS97-085-3
bottom of the core PS97-085-3_TC
bottom of the core PS97-089-1
bottom of the core PS97-089-1_TC
bottom of the core PS97-092-1
bottom of the core PS97-093-2
bottom of the core PS97-128-2


In [20]:
clean_df.head()

0  1  2  3  4  5  6  7  8  9  ...  2044  2045  2046  2047  \
composite_id                                    ...                           
PS75-054-1_00010  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
PS75-054-1_00020  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
PS75-054-1_00030  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
PS75-054-1_00040  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
PS75-054-1_00050  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   

                     cps        core  composite_depth_mm  section_depth_mm  \
composite_id                                                                 
PS75-054-1_00010  139498  PS75-054-1                  10                10   
PS75-054-1_00020  137079  PS75-054-1                  20                20   
PS75-054-1_00030  140370  PS75-054-1                  30                30   
PS75-054-1_00040  138782  PS75-054-1                  40                40   
PS75-054-1_00050  137974  PS75-054-1                  50                50   

                                                           filename  section  
composite_id                                                                  
PS75-054-1_00010  PS75-054-1_0000   10.0mm  10s  10kV 150uA No-F...        0  
PS75-054-1_00020  PS75-054-1_0000   20.0mm  10s  10kV 150uA No-F...        0  
PS75-054-1_00030  PS75-054-1_0000   30.0mm  10s  10kV 150uA No-F...        0  
PS75-054-1_00040  PS75-054-1_0000   40.0mm  10s  10kV 150uA No-F...        0  
PS75-054-1_00050  PS75-054-1_0000   50.0mm  10s  10kV 150uA No-F...        0  

[5 rows x 2054 columns]

## Build bulk chemistry dataset

In [52]:
with open('results/corewithbulks.txt', 'w') as f:
    for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
        text = 'Core {} has: {}'.format(
            core, 
            glob.glob('data/Bulk chem/{}*'.format(core)))
        print(text)
        print(text, file=f)

Core PS97-027-2 has: ['data/Bulk chem/PS97-027-2_bulk CaCO3.xlsx']
Core PS97-046-4 has: ['data/Bulk chem/PS97-046-4_bulk CaCO3.xlsx']
Core PS97-052-4 has: ['data/Bulk chem/PS97-052-4_bulk CaCO3.xlsx']
Core PS97-053-2 has: ['data/Bulk chem/PS97-053-2_bulk CaCO3.xlsx']
Core PS97-078-1 has: ['data/Bulk chem/PS97-078-1_bulk CaCO3.xlsx']
Core PS97-079-2 has: ['data/Bulk chem/PS97-079-2_bulk CaCO3.xlsx']
Core PS97-080-1 has: ['data/Bulk chem/PS97-080-1_bulk CaCO3.xlsx']
Core PS97-083-2 has: ['data/Bulk chem/PS97-083-2_bulk CaCO3.xlsx']
Core PS97-084-1 has: ['data/Bulk chem/PS97-084-1_bulk CaCO3.xlsx', 'data/Bulk chem/PS97-084-1 TC_bulk CaCO3.xlsx']
Core PS97-085-3 has: ['data/Bulk chem/PS97-085-3_bulk CaCO3.xlsx', 'data/Bulk chem/PS97-085-3 TC_bulk CaCO3.xlsx']
Core PS97-089-1 has: ['data/Bulk chem/PS97-089-1_bulk CaCO3.xlsx', 'data/Bulk chem/PS97-089-1 TC_bulk CaCO3.xlsx']
Core PS97-092-1 has: ['data/Bulk chem/PS97-092-1_bulk CaCO3.xlsx']
Core PS97-093-2 has: ['data/Bulk chem/PS97-093-2_bul

Let's deal with the bulk files having "CaCO3.xlsx" end first. Later I do the triger cores.

In [54]:
for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
    text = 'Core {} has: {}'.format(
        core, 
        glob.glob('data/Bulk chem/{}_bulk CaCO3.xlsx'.format(core)))
    print(text)

Core PS97-027-2 has: ['data/Bulk chem/PS97-027-2_bulk CaCO3.xlsx']
Core PS97-046-4 has: ['data/Bulk chem/PS97-046-4_bulk CaCO3.xlsx']
Core PS97-052-4 has: ['data/Bulk chem/PS97-052-4_bulk CaCO3.xlsx']
Core PS97-053-2 has: ['data/Bulk chem/PS97-053-2_bulk CaCO3.xlsx']
Core PS97-078-1 has: ['data/Bulk chem/PS97-078-1_bulk CaCO3.xlsx']
Core PS97-079-2 has: ['data/Bulk chem/PS97-079-2_bulk CaCO3.xlsx']
Core PS97-080-1 has: ['data/Bulk chem/PS97-080-1_bulk CaCO3.xlsx']
Core PS97-083-2 has: ['data/Bulk chem/PS97-083-2_bulk CaCO3.xlsx']
Core PS97-084-1 has: ['data/Bulk chem/PS97-084-1_bulk CaCO3.xlsx']
Core PS97-085-3 has: ['data/Bulk chem/PS97-085-3_bulk CaCO3.xlsx']
Core PS97-089-1 has: ['data/Bulk chem/PS97-089-1_bulk CaCO3.xlsx']
Core PS97-092-1 has: ['data/Bulk chem/PS97-092-1_bulk CaCO3.xlsx']
Core PS97-093-2 has: ['data/Bulk chem/PS97-093-2_bulk CaCO3.xlsx']
Core PS97-128-2 has: ['data/Bulk chem/PS97-128-2_bulk CaCO3.xlsx']
Core PS75-054-1 has: ['data/Bulk chem/PS75-054-1_bulk CaCO3.xl

In [56]:
bulk_df = pd.DataFrame()
for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
    dir = glob.glob('data/Bulk chem/{}_bulk CaCO3.xlsx'.format(core))[0]
    bulk_df = pd.concat(
        [bulk_df, 
         pd.read_excel(dir)],
        join='outer', axis=0)
bulk_df

depth cm      TC %     TOC %    CaCO3 %  Unnamed: 4  Unnamed: 5  \
0        10.0  7.411925  0.123642  60.733260         NaN         NaN   
1        20.0  3.338068  0.159575  26.486376         NaN         NaN   
2        30.0  1.142647  0.156978   8.213584         NaN         NaN   
3        40.0  0.933463  0.198723   6.122586         NaN         NaN   
4        50.0  0.517007  0.260630   2.136389         NaN         NaN   
..        ...       ...       ...        ...         ...         ...   
170    1740.0  0.197832  0.166828   0.258354         NaN         NaN   
171    1750.0  0.217313  0.192214   0.209152         NaN         NaN   
172    1760.0  3.293107  0.121690  26.427417         NaN         NaN   
173    1770.0  5.383270  0.120147  43.857604         NaN         NaN   
174    1780.0  3.830366  0.155866  30.619614         NaN         NaN   

     Unnamed: 6  Unnamed: 7  Unnamed: 8  
0           NaN         NaN         NaN  
1           NaN         NaN         NaN  
2           NaN         NaN         NaN  
3           NaN         NaN         NaN  
4           NaN         NaN         NaN  
..          ...         ...         ...  
170         NaN         NaN         NaN  
171         NaN         NaN         NaN  
172         NaN         NaN         NaN  
173         NaN         NaN         NaN  
174         NaN         NaN         NaN  

[1798 rows x 9 columns]

In [57]:
for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
    dir = glob.glob('data/Bulk chem/{}_bulk CaCO3.xlsx'.format(core))[0]
    print('{}: {}'.format(core, pd.read_excel(dir).columns))

PS97-027-2: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-046-4: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-052-4: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-053-2: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-078-1: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-079-2: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-080-1: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-083-2: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-084-1: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-085-3: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %', 'Unnamed: 4', 'Unnamed: 5'], dtype='object')
PS97-089-1: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-092-1: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], dtype='object')
PS97-093-2: Index(['depth cm', 'TC %', 'TOC %', 'CaCO3 %'], 

The Unhamed columns should be some manual marks, which are not relevant in our interest. The desired columns, 'depth cm', 'TC %', 'TOC %', 'CaCO3 %', are named consistently, so I'll just delete those Unhamed columns.

In [21]:
bulk_df = pd.DataFrame()
core_list = []
for core in xl_df.Station[~xl_df.Station.isin(pre_df.core.unique())]:
    X = pd.read_excel(
        glob.glob('data/Bulk chem/{}_bulk CaCO3.xlsx'.format(core))[0], 
        usecols=[0, 1, 2, 3])
    bulk_df = pd.concat(
        [bulk_df, X],
        join='outer', axis=0)
    core_list = np.hstack([core_list, [core for _ in range(len(X))]])

bulk_df['core'] = core_list
bulk_df

depth cm      TC %     TOC %    CaCO3 %        core
0        10.0  7.411925  0.123642  60.733260  PS97-027-2
1        20.0  3.338068  0.159575  26.486376  PS97-027-2
2        30.0  1.142647  0.156978   8.213584  PS97-027-2
3        40.0  0.933463  0.198723   6.122586  PS97-027-2
4        50.0  0.517007  0.260630   2.136389  PS97-027-2
..        ...       ...       ...        ...         ...
170    1740.0  0.197832  0.166828   0.258354  PS75-095-5
171    1750.0  0.217313  0.192214   0.209152  PS75-095-5
172    1760.0  3.293107  0.121690  26.427417  PS75-095-5
173    1770.0  5.383270  0.120147  43.857604  PS75-095-5
174    1780.0  3.830366  0.155866  30.619614  PS75-095-5

[1620 rows x 5 columns]

In [22]:
bulk_df.isna().any()

depth cm    False
TC %        False
TOC %       False
CaCO3 %     False
core        False
dtype: bool

### Add the measurments of Triger cores

In [26]:
bulk_tc_df = pd.DataFrame()
core_tc_list = []
# These are the cores having TC as subfolders
for core in ['PS97-084-1', 'PS97-085-3', 'PS97-089-1', 'PS75-093-1', 'PS75-083-1']:
    X = pd.read_excel(
        glob.glob('data/Bulk chem/{} TC_bulk CaCO3.xlsx'.format(core))[0], 
        usecols=[0, 1, 2, 3])
    bulk_tc_df = pd.concat(
        [bulk_tc_df, X],
        join='outer', axis=0)
    core_tc_list = np.hstack([core_tc_list, [core+'_TC' for _ in range(len(X))]])

bulk_tc_df['core'] = core_tc_list
print(bulk_tc_df)
print(bulk_tc_df.shape)

   depth cm      TC %     TOC %    CaCO3 %           core
0        10  1.117363  0.207447   7.582325  PS97-084-1_TC
1        20  1.011192  0.230752   6.503406  PS97-084-1_TC
2        30  0.713651  0.269405   3.701905  PS97-084-1_TC
3        40  1.737001  0.229905  12.558624  PS97-084-1_TC
4        50  1.164319  0.205090   7.993254  PS97-084-1_TC
5        60  1.409437  0.192600  10.139910  PS97-084-1_TC
6        70  1.057188  0.204069   7.109038  PS97-084-1_TC
7        80  0.932773  0.230438   5.852561  PS97-084-1_TC
0        10  2.809865  0.063382  22.886439  PS97-085-3_TC
1        20  2.720691  0.082306  21.985658  PS97-085-3_TC
2        30  2.537987  0.068949  20.574495  PS97-085-3_TC
3        40  2.575967  0.075634  20.835278  PS97-085-3_TC
4        50  2.430658  0.067202  19.694677  PS97-085-3_TC
5        60  2.638385  0.066749  21.429447  PS97-085-3_TC
6        70  2.899329  0.067848  23.594730  PS97-085-3_TC
7        80  2.642680  0.077441  21.376141  PS97-085-3_TC
0        10  1

In [27]:
print(bulk_df.shape)
bulk_df = pd.concat([bulk_df, bulk_tc_df], join='outer', axis=0)
print(bulk_df.shape)

(1620, 5)
(1657, 5)


In [28]:
bulk_df.isna().any()

depth cm    False
TC %        False
TOC %       False
CaCO3 %     False
core        False
dtype: bool

### Check duplicates

In [35]:
composite_id = []
for _, row in bulk_df.iterrows():
    composite_id.append('{}_{:05}'.format(row['core'], row['depth cm']))

any(composite_id.count(element) > 1 for element in composite_id)

False

### Transform depth_cm to mid_depth_mm

In [39]:
for _, row in bulk_df.iterrows():
    if str(row['depth cm'])[-3] == 5:
        print(row)

Okay, no depth ends with 0.5 cm, so they all need to be plus 0.5 cm to be middle depth.

In [43]:
bulk_df['mid_depth_mm'] = bulk_df['depth cm'].astype(int)*10 + 5
bulk_df.drop('depth cm', axis=1, inplace=True)
bulk_df.head()

TC %     TOC %    CaCO3 %        core  mid_depth_mm
0  7.411925  0.123642  60.733260  PS97-027-2           105
1  3.338068  0.159575  26.486376  PS97-027-2           205
2  1.142647  0.156978   8.213584  PS97-027-2           305
3  0.933463  0.198723   6.122586  PS97-027-2           405
4  0.517007  0.260630   2.136389  PS97-027-2           505

In [52]:
# This is to be consistent to the previous data
bulk_df.columns = ['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm']

## Buld merged spe and bulk dataset

In [84]:
mask_c = spe_df.columns[:2048]  # only the channels
merge_df = pd.DataFrame()

for _, row in bulk_df.iterrows():
    mid = row['mid_depth_mm']
    core = row['core']
    
    # get the spe in 10 mm interval
    mask_r = ((spe_df.composite_depth_mm >= (mid-5)) & 
              (spe_df.composite_depth_mm <= (mid+5)) & 
              (spe_df.core == core))
    merge_df = pd.concat(
        [merge_df, spe_df.loc[mask_r, mask_c].apply(np.mean, axis = 0).append(row)],
        axis = 1
    )
    
merge_df = merge_df.T.reset_index(drop = True)

In [88]:
print(merge_df)
print(merge_df.info())

        0    1    2    3    4    5    6    7    8    9  ... 2043 2044 2045  \
0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1652  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1653  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1654  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1655  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1656  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

     2046 2047       TC%      TOC%     CaCO3%           core mi

In [97]:
_ = merge_df.loc[merge_df.isna().any(axis = 1), merge_df.columns[-6:]]
_.to_csv('results/merge_NAs.csv')
print(_.shape)

(129, 6)


After checking the csv, I find these are the data points that having bulk chemistry measurements but no XRF. Later I will only use the rows without NAs to merge with the previous data. If there are some bulk measurements missing but we still want to keep the data points (e.g. having TOC but no CaCO3), when filtering data, beter use `merge_df[~merge_df.iloc[:, :2048].isna().any(axis = 1)]` to address the missing XRF issue specificly.

# Merge the new datasets with the previous datasets

## Spetral dataset

In [48]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/spe_dataset_20201215.csv', index_col=0),
           pd.read_csv('data/spe_PS75-056-1_20210601.csv', index_col=0),
           clean_df]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns[2048:])

(37823, 2054)
Index(['cps', 'core', 'composite_depth_mm', 'section_depth_mm', 'filename',
       'section'],
      dtype='object')
(40053, 2054)
Index(['cps', 'core', 'composite_depth_mm', 'section_depth_mm', 'filename',
       'section'],
      dtype='object')
(57240, 2054)
Index(['cps', 'core', 'composite_depth_mm', 'section_depth_mm', 'filename',
       'section'],
      dtype='object')


The columns are consistent because no new column show up during concating. The increase of row amount seems to be reasonable. The spe dataset already has its composite_id as the index so no need for reseting the index.

In [50]:
c_df.to_csv('data/spe_dataset_{}.csv'.format(date))

## Bulk chemistry

In [53]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/bulk_dataset_20201215.csv', index_col=0),
           pd.read_csv('data/bulk_PS75-056-1_20210601.csv', index_col=0),
           bulk_df]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns)

(742, 5)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(845, 5)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(2502, 5)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')


In [67]:
c_df.reset_index(drop=True, inplace=True)
c_df[c_df.mid_depth_mm.isna()]

Empty DataFrame
Columns: [TC%, TOC%, CaCO3%, core, mid_depth_mm]
Index: []

Le'ts check if the value changes after changing to integer

In [68]:
c_df[c_df.mid_depth_mm.astype(int) != c_df.mid_depth_mm]

TC%  TOC%  CaCO3%        core  mid_depth_mm
823  NaN   NaN  24.422  PS75-056-1        8050.0

Weird...the one filtered out due to its difference between making integer actually doesn't have difference... int(8050.0) should and is equal to 8050...

In [72]:
print(c_df.iloc[823, -1])
print(int(c_df.iloc[823, -1]))

8050.000000000001
8050


I see...I can ignore it without worry.

In [77]:
c_df.mid_depth_mm.max()

22305.0

In [80]:
composite_id = []
c_df.mid_depth_mm = c_df.mid_depth_mm.astype(int)
for _, row in c_df.iterrows():
    composite_id.append('{}_{:05}'.format(row.core, row.mid_depth_mm))

c_df.index = composite_id
c_df.head()

TC%      TOC%     CaCO3%        core  mid_depth_mm
SO264-64-1_00115  2.542079  0.394127  17.898887  SO264-64-1           115
SO264-64-1_00215  2.247150  0.611208  13.632300  SO264-64-1           215
SO264-64-1_00305  0.710588  0.523402   1.559822  SO264-64-1           305
SO264-64-1_01015  0.562171  0.472551   0.746802  SO264-64-1          1015
SO264-64-1_01815  0.578167  0.312852   2.210866  SO264-64-1          1815

In [81]:
# Just check again although it should be checked everytime building dataset
c_df[c_df.index.duplicated()]

Empty DataFrame
Columns: [TC%, TOC%, CaCO3%, core, mid_depth_mm]
Index: []

In [82]:
c_df.to_csv('data/bulk_dataset_{}.csv'.format(date))

## spe+bulk

In [98]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/spe+bulk_dataset_20201215.csv', index_col=0),
           pd.read_csv('data/spe+bulk_PS75-056-1_20210601.csv', index_col=0),
           merge_df[~merge_df.isna().any(axis = 1)]]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns[2048:])

(699, 2053)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(802, 2053)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(2330, 2053)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')


In [99]:
c_df.to_csv('data/spe+bulk_dataset_{}.csv'.format(date))

# Include SO202-37-2_rescan to the database
This core is left over so I include it into the database. The output datasets are still using the output date 20210818 since it's just a small update.

In [2]:
merge_cores = pd.read_csv('data/spe+bulk_dataset_20210818.csv', usecols=['core'], squeeze=True)
spe_cores = pd.read_csv('data/spe_dataset_20210818.csv', usecols=['core'], squeeze=True)
bulk_cores = pd.read_csv('data/bulk_dataset_20210818.csv', usecols=['core'], squeeze=True)

In [6]:
for core in ['SO202-37-2', 'PS75-056-1']:
    for data in  [merge_cores, spe_cores, bulk_cores]:
        print(core in data.unique())

False
False
False
True
True
True


## Build spectral dataset

In [28]:
file_name = []
spe_all = []
depth_all = []
cps_all = []
core_all = []
s_depth_all = []

core = 'SO202-37-2_re'
# only read the 10kV which having better signal to the light elements
# the dir is specified to the SO202-37-2 rescan
spe_dir = glob.glob('data/SO202-37-2_2021rescan_937-1138cm/Run 1 at  10kV_937-1138cm/*.spe')

# make sure the order follows the depthes in filename
spe_dir.sort()
print('{} has {} data points.'.format(core, len(spe_dir)))

for spe in spe_dir:
    check_depth = spe.split()[3].split('_')[-1]

    # 5 and 6 digis means in mm
    if len(check_depth) >= 5:
        start_depth = int(check_depth)
    # 3 and 4 digit means in cm, needs to be multipled to be mm
    elif len(check_depth) >= 3:
        start_depth = int(check_depth) * 10

    file_name.append(spe.split('/')[-1])

    with open(spe, 'r') as f:
        content = []
        lines = f.readlines()
        for line in lines[49:]:
            content = np.hstack((content, line.split()))
        
        section_depth = int(lines[13][:-3])
        spe_all.append(content.astype(int))
        cps_all.append(int(lines[28]))
        core_all.append(core)
        s_depth_all.append(section_depth)
        depth_all.append(section_depth + start_depth)

print('core {} is done.'.format(core))

SO202-37-2_re has 201 data points.
core SO202-37-2_re is done.


In [29]:
spe_df = pd.DataFrame(spe_all, columns = [str(_) for _ in range(2048)])
spe_df['cps'] = cps_all
spe_df['core'] = core_all
spe_df['composite_depth_mm'] = depth_all
spe_df['section_depth_mm'] = s_depth_all
spe_df['filename'] = file_name

spe_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Columns: 2053 entries, 0 to filename
dtypes: int64(2051), object(2)
memory usage: 3.1+ MB


In [30]:
spe_df[spe_df.isna().any(axis=1)]

Empty DataFrame
Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]
Index: []

[0 rows x 2053 columns]

### Build composite_id

In [31]:
composite_id = []
for core, depth in zip(spe_df.core, spe_df.composite_depth_mm):
    composite_id.append('{}_{:05}'.format(core, depth))
    
spe_df['composite_id'] = composite_id

### Drop duplicates

In [32]:
clean_df = spe_df.drop_duplicates('composite_id', keep = 'last')
len(clean_df)

201

The length is same as the spe_df, so there is no duplicate. But due to the consistency of the codes, I use clean_df instead of spe_df.

### Build section no.

In [33]:
clean_df = clean_df.set_index('composite_id')

section_all = []
# make sure the order follows the core and composite depth
clean_df.sort_values(by = 'composite_id', axis = 0, inplace = True)

for core in np.unique(clean_df.core):    
    # I assume every core scanned from section 0 so the first section in the core 
    # is marked as section 0. The deeper the larger number
    section = 0
    X = clean_df.loc[clean_df.core == core, 'section_depth_mm']
    for i in range(len(X)):
        section_all.append(section)
        try:
            # when section changes, the section depth should be reset to smaller number
            if X[i] > X[i + 1]:
                section += 1
        except IndexError:
            print('bottom of the core {}'.format(core))
        
clean_df['section'] = section_all

bottom of the core SO202-37-2_re


In [34]:
clean_df.head()

0  1  2  3  4  5  6  7  8  9  ...  2044  2045  2046  \
composite_id                                       ...                     
SO202-37-2_re_09380  0  0  0  0  0  0  0  0  0  0  ...     0     0     0   
SO202-37-2_re_09390  0  0  0  0  0  0  0  0  0  0  ...     0     0     0   
SO202-37-2_re_09400  0  0  0  0  0  0  0  0  0  0  ...     0     0     0   
SO202-37-2_re_09410  0  0  0  0  0  0  0  0  0  0  ...     0     0     0   
SO202-37-2_re_09420  0  0  0  0  0  0  0  0  0  0  ...     0     0     0   

                     2047     cps           core  composite_depth_mm  \
composite_id                                                           
SO202-37-2_re_09380     0  164784  SO202-37-2_re                9380   
SO202-37-2_re_09390     0  193043  SO202-37-2_re                9390   
SO202-37-2_re_09400     0  189160  SO202-37-2_re                9400   
SO202-37-2_re_09410     0  190785  SO202-37-2_re                9410   
SO202-37-2_re_09420     0  193711  SO202-37-2_re                9420   

                     section_depth_mm  \
composite_id                            
SO202-37-2_re_09380                10   
SO202-37-2_re_09390                20   
SO202-37-2_re_09400                30   
SO202-37-2_re_09410                40   
SO202-37-2_re_09420                50   

                                                              filename  \
composite_id                                                             
SO202-37-2_re_09380  SO202-37-2_0937   10.0mm  10s  10kV 150uA No-F...   
SO202-37-2_re_09390  SO202-37-2_0937   20.0mm  10s  10kV 150uA No-F...   
SO202-37-2_re_09400  SO202-37-2_0937   30.0mm  10s  10kV 150uA No-F...   
SO202-37-2_re_09410  SO202-37-2_0937   40.0mm  10s  10kV 150uA No-F...   
SO202-37-2_re_09420  SO202-37-2_0937   50.0mm  10s  10kV 150uA No-F...   

                     section  
composite_id                  
SO202-37-2_re_09380        0  
SO202-37-2_re_09390        0  
SO202-37-2_re_09400        0  
SO202-37-2_re_09410        0  
SO202-37-2_re_09420        0  

[5 rows x 2054 columns]

## Build bulk dataset

In [20]:
pd.read_excel('data/Bulk chem/SO202-37-2_CaCO3 TOC_940-1130cm.xlsx')

37-2 Depth cm  37-2 carbonate %  37-2 TOC %
0             940              67.9    0.208847
1             950              75.6    0.204656
2             960              85.3    0.146924
3             970              82.7    0.200416
4             980              74.5    0.204804
5             990              68.6    0.334823
6            1000              80.2    0.198436
7            1010              84.8    0.156819
8            1020              77.7    0.188815
9            1030              87.7    0.166428
10           1040              91.3    0.120992
11           1050              84.6    0.121362
12           1060              77.4    0.269102
13           1070              62.0    0.174479
14           1080              50.4    0.366214
15           1090              87.1    0.222168
16           1100              86.8    0.216847
17           1110              91.0    0.244540
18           1120              92.3    0.296126
19           1130              93.2    0.194526

In [22]:
bulk_df = pd.read_excel('data/Bulk chem/SO202-37-2_CaCO3 TOC_940-1130cm.xlsx')
bulk_df.columns = ['depth cm', 'CaCO3%', 'TOC%']
bulk_df['core'] = ['SO202-37-2_re' for _ in range(len(bulk_df))]
bulk_df['mid_depth_mm'] = bulk_df['depth cm']*10 + 5
bulk_df.drop('depth cm', axis=1, inplace=True)
bulk_df

CaCO3%      TOC%           core  mid_depth_mm
0     67.9  0.208847  SO202-37-2_re          9405
1     75.6  0.204656  SO202-37-2_re          9505
2     85.3  0.146924  SO202-37-2_re          9605
3     82.7  0.200416  SO202-37-2_re          9705
4     74.5  0.204804  SO202-37-2_re          9805
5     68.6  0.334823  SO202-37-2_re          9905
6     80.2  0.198436  SO202-37-2_re         10005
7     84.8  0.156819  SO202-37-2_re         10105
8     77.7  0.188815  SO202-37-2_re         10205
9     87.7  0.166428  SO202-37-2_re         10305
10    91.3  0.120992  SO202-37-2_re         10405
11    84.6  0.121362  SO202-37-2_re         10505
12    77.4  0.269102  SO202-37-2_re         10605
13    62.0  0.174479  SO202-37-2_re         10705
14    50.4  0.366214  SO202-37-2_re         10805
15    87.1  0.222168  SO202-37-2_re         10905
16    86.8  0.216847  SO202-37-2_re         11005
17    91.0  0.244540  SO202-37-2_re         11105
18    92.3  0.296126  SO202-37-2_re         11205
19    93.2  0.194526  SO202-37-2_re         11305

In [47]:
composite_id = []
for _, row in bulk_df.iterrows():
    composite_id.append('{}_{:05}'.format(row.core, row.mid_depth_mm))

bulk_df.index = composite_id
bulk_df.head()

CaCO3%      TOC%           core  mid_depth_mm
SO202-37-2_re_09405    67.9  0.208847  SO202-37-2_re          9405
SO202-37-2_re_09505    75.6  0.204656  SO202-37-2_re          9505
SO202-37-2_re_09605    85.3  0.146924  SO202-37-2_re          9605
SO202-37-2_re_09705    82.7  0.200416  SO202-37-2_re          9705
SO202-37-2_re_09805    74.5  0.204804  SO202-37-2_re          9805

### Check duplicates

In [48]:
bulk_df[bulk_df.index.duplicated()]

Empty DataFrame
Columns: [CaCO3%, TOC%, core, mid_depth_mm]
Index: []

## Build merged spe and bulk dataset

In [44]:
mask_c = spe_df.columns[:2048]  # only the channels
merge_df = pd.DataFrame()

for _, row in bulk_df.iterrows():
    mid = row['mid_depth_mm']
    core = row['core']
    
    # get the spe in 10 mm interval
    mask_r = ((spe_df.composite_depth_mm >= (mid-5)) & 
              (spe_df.composite_depth_mm <= (mid+5)) & 
              (spe_df.core == core))
    merge_df = pd.concat(
        [merge_df, spe_df.loc[mask_r, mask_c].apply(np.mean, axis = 0).append(row)],
        axis = 1
    )
    
merge_df = merge_df.T.reset_index(drop = True)

In [45]:
print(merge_df.info())
print(merge_df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Columns: 2052 entries, 0 to mid_depth_mm
dtypes: object(2052)
memory usage: 320.8+ KB
None
     0    1    2    3    4    5    6    7    8    9  ... 2042 2043 2044 2045  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

  2046 2047 CaCO3%      TOC%           core mid_depth_mm  
0  0.0  0.0   67.9  0.208847  SO202-37-2_re         9405  
1  0.0  0.0   75.6  0.204656  SO202-37-2_re         9505  
2  0.0  0.0   85.3  0.146924  SO202-37-2_re         9605  
3  0.0  0.0   82.7  0.200416  SO202-37-2_re         9705  
4  0.0  0.0   74.5  0.204804  SO202-37-2_re         9805  

[5 

In [46]:
merge_df.loc[merge_df.isna().any(axis = 1), merge_df.columns[-6:]]

Empty DataFrame
Columns: [2046, 2047, CaCO3%, TOC%, core, mid_depth_mm]
Index: []

## Update the 20210818 database

### Spetral dataset

In [49]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/spe_dataset_20210818.csv', index_col=0),
           clean_df]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns[2048:])
c_df.head()

(57441, 2054)
Index(['cps', 'core', 'composite_depth_mm', 'section_depth_mm', 'filename',
       'section'],
      dtype='object')
(57642, 2054)
Index(['cps', 'core', 'composite_depth_mm', 'section_depth_mm', 'filename',
       'section'],
      dtype='object')


0  1  2  3  4  5  6  7  8  9  ...  2044  2045  2046  2047  \
composite_id                                    ...                           
SO264-09-2_00010  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
SO264-09-2_00020  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
SO264-09-2_00030  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
SO264-09-2_00040  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   
SO264-09-2_00050  0  0  0  0  0  0  0  0  0  0  ...     0     0     0     0   

                     cps        core  composite_depth_mm  section_depth_mm  \
composite_id                                                                 
SO264-09-2_00010  174740  SO264-09-2                  10                10   
SO264-09-2_00020  176896  SO264-09-2                  20                20   
SO264-09-2_00030  183148  SO264-09-2                  30                30   
SO264-09-2_00040  185151  SO264-09-2                  40                40   
SO264-09-2_00050  186255  SO264-09-2                  50                50   

                                                           filename  section  
composite_id                                                                  
SO264-09-2_00010  SO264-09-2_0000   10.0mm  10s  10kV 150uA No-F...        0  
SO264-09-2_00020  SO264-09-2_0000   20.0mm  10s  10kV 150uA No-F...        0  
SO264-09-2_00030  SO264-09-2_0000   30.0mm  10s  10kV 150uA No-F...        0  
SO264-09-2_00040  SO264-09-2_0000   40.0mm  10s  10kV 150uA No-F...        0  
SO264-09-2_00050  SO264-09-2_0000   50.0mm  10s  10kV 150uA No-F...        0  

[5 rows x 2054 columns]

In [40]:
c_df.to_csv('data/spe_dataset_20210818.csv')

In [62]:
pd.read_csv('data/spe_dataset_20210818.csv', index_col=0).info()

<class 'pandas.core.frame.DataFrame'>
Index: 57441 entries, SO264-09-2_00010 to SO202-37-2_re_11380
Columns: 2054 entries, 0 to section
dtypes: int64(2052), object(2)
memory usage: 900.6+ MB


### Bulk chemistry

In [56]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/bulk_dataset_20210818.csv', index_col=0),
           bulk_df]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns)

(2502, 5)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(2522, 5)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')


In [57]:
c_df

TC%      TOC%     CaCO3%           core  \
SO264-64-1_00115     2.542079  0.394127  17.898887     SO264-64-1   
SO264-64-1_00215     2.247150  0.611208  13.632300     SO264-64-1   
SO264-64-1_00305     0.710588  0.523402   1.559822     SO264-64-1   
SO264-64-1_01015     0.562171  0.472551   0.746802     SO264-64-1   
SO264-64-1_01815     0.578167  0.312852   2.210866     SO264-64-1   
...                       ...       ...        ...            ...   
SO202-37-2_re_10905       NaN  0.222168  87.100000  SO202-37-2_re   
SO202-37-2_re_11005       NaN  0.216847  86.800000  SO202-37-2_re   
SO202-37-2_re_11105       NaN  0.244540  91.000000  SO202-37-2_re   
SO202-37-2_re_11205       NaN  0.296126  92.300000  SO202-37-2_re   
SO202-37-2_re_11305       NaN  0.194526  93.200000  SO202-37-2_re   

                     mid_depth_mm  
SO264-64-1_00115              115  
SO264-64-1_00215              215  
SO264-64-1_00305              305  
SO264-64-1_01015             1015  
SO264-64-1_01815             1815  
...                           ...  
SO202-37-2_re_10905         10905  
SO202-37-2_re_11005         11005  
SO202-37-2_re_11105         11105  
SO202-37-2_re_11205         11205  
SO202-37-2_re_11305         11305  

[2522 rows x 5 columns]

In [58]:
c_df.to_csv('data/bulk_dataset_20210818.csv')

### spe+bulk

In [59]:
c_df = pd.DataFrame()
for df in [pd.read_csv('data/spe+bulk_dataset_20210818.csv', index_col=0),
           merge_df]:
    c_df = pd.concat([c_df, df], join='outer', axis=0)
    print(c_df.shape)
    print(c_df.columns[2048:])

(2330, 2053)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')
(2350, 2053)
Index(['TC%', 'TOC%', 'CaCO3%', 'core', 'mid_depth_mm'], dtype='object')


In [60]:
c_df

0    1    2    3    4    5    6    7    8    9  ... 2043 2044 2045 2046  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
15  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
16  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
17  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
18  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
19  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   2047       TC%      TOC%     CaCO3%           core mid_depth_mm  
0   0.0  2.542079  0.394127  17.898887     SO264-64-1        115.0  
1   0.0  2.247150  0.611208    13.6323     SO264-64-1        215.0  
2   0.0  0.710588  0.523402   1.559822     SO264-64-1        305.0  
3   0.0  0.562171  0.472551   0.746802     SO264-64-1       1015.0  
4   0.0  0.578167  0.312852   2.210866     SO264-64-1       1815.0  
..  ...       ...       ...        ...            ...          ...  
15  0.0       NaN  0.222168       87.1  SO202-37-2_re        10905  
16  0.0       NaN  0.216847       86.8  SO202-37-2_re        11005  
17  0.0       NaN   0.24454       91.0  SO202-37-2_re        11105  
18  0.0       NaN  0.296126       92.3  SO202-37-2_re        11205  
19  0.0       NaN  0.194526       93.2  SO202-37-2_re        11305  

[2350 rows x 2053 columns]

In [61]:
c_df.to_csv('data/spe+bulk_dataset_20210818.csv')

# Update the bulk measurement of PS75-056-1
This core is included in dataset_20210818, but missing TOC measurement in the bulk chemistry (the old data: PS75-56_bulkCaCO3 (%)). We update the bulk measurement by PS75-056-1_bulk CaCO3.xlsx.

In [4]:
pd.read_excel('data/Bulk chem/PS75-056-1_bulk CaCO3.xlsx', usecols=[0, 1, 2, 3])

depth cm      TC %     TOC %  CaCO3 %
0         1.0  9.596000  0.150502   78.870
1         5.0  9.457088  0.129249   77.889
2        15.0  9.124211  0.103419   75.332
3        25.0  9.276967  0.095119   76.673
4        35.0  8.877704  0.117070   73.165
..        ...       ...       ...      ...
98      975.0  3.137845  0.268411   24.091
99      985.0  3.434429  0.267635   26.568
100     995.0  3.583974  0.239806   28.045
101    1005.0  3.125558  0.182222   24.707
102    1015.0  3.025567  0.203551   23.696

[103 rows x 4 columns]

In [23]:
bulk_df = pd.read_excel('data/Bulk chem/PS75-056-1_bulk CaCO3.xlsx', 
                        usecols=[0, 1, 2, 3])
bulk_df.columns = ['depth cm', 'TC%', 'TOC%', 'CaCO3%']
bulk_df['core'] = ['PS75-056-1' for _ in range(len(bulk_df))]
bulk_df['mid_depth_mm'] = bulk_df['depth cm'].astype(int)*10 + 5
bulk_df.drop('depth cm', axis=1, inplace=True)
bulk_df

TC%      TOC%  CaCO3%        core  mid_depth_mm
0    9.596000  0.150502  78.870  PS75-056-1            15
1    9.457088  0.129249  77.889  PS75-056-1            55
2    9.124211  0.103419  75.332  PS75-056-1           155
3    9.276967  0.095119  76.673  PS75-056-1           255
4    8.877704  0.117070  73.165  PS75-056-1           355
..        ...       ...     ...         ...           ...
98   3.137845  0.268411  24.091  PS75-056-1          9755
99   3.434429  0.267635  26.568  PS75-056-1          9855
100  3.583974  0.239806  28.045  PS75-056-1          9945
101  3.125558  0.182222  24.707  PS75-056-1         10055
102  3.025567  0.203551  23.696  PS75-056-1         10155

[103 rows x 5 columns]

### Assign composite_id and check duplicates

In [24]:
composite_id = []
for _, row in bulk_df.iterrows():
    composite_id.append('{}_{:05}'.format(row.core, row.mid_depth_mm))

bulk_df.index = composite_id
bulk_df[bulk_df.index.duplicated()]

Empty DataFrame
Columns: [TC%, TOC%, CaCO3%, core, mid_depth_mm]
Index: []

## Build merged spe and bulk dataset
I adopt the spe dataset since this core's spectra are already included.

In [11]:
spe_df = pd.read_csv('data/spe_dataset_20210818.csv', index_col=0)

In [25]:
mask_c = spe_df.columns[:2048]  # only the channels
merge_df = pd.DataFrame()

for _, row in bulk_df.iterrows():
    mid = row['mid_depth_mm']
    core = row['core']
    
    # get the spe in 10 mm interval
    mask_r = ((spe_df.composite_depth_mm >= (mid-5)) & 
              (spe_df.composite_depth_mm <= (mid+5)) & 
              (spe_df.core == core))
    merge_df = pd.concat(
        [merge_df, spe_df.loc[mask_r, mask_c].apply(np.mean, axis = 0).append(row)],
        axis = 1
    )
    
merge_df = merge_df.T.reset_index(drop = True)

In [26]:
print(merge_df.info())
print(merge_df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Columns: 2053 entries, 0 to mid_depth_mm
dtypes: object(2053)
memory usage: 1.6+ MB
None
     0    1    2    3    4    5    6    7    8    9  ... 2043 2044 2045 2046  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

  2047       TC%      TOC%  CaCO3%        core mid_depth_mm  
0  0.0     9.596  0.150502   78.87  PS75-056-1           15  
1  0.0  9.457088  0.129249  77.889  PS75-056-1           55  
2  0.0  9.124211  0.103419  75.332  PS75-056-1          155  
3  0.0  9.276967  0.095119  76.673  PS75-056-1          255  
4  0.0  8.877704   0.11707  73.165  PS75-056-1  

In [27]:
merge_df.loc[merge_df.isna().any(axis = 1), merge_df.columns[-6:]]

Empty DataFrame
Columns: [2047, TC%, TOC%, CaCO3%, core, mid_depth_mm]
Index: []

Good, all bulk measurements can find relevant XRF spectra.

## Update the 20210818 database again
This time we produce another datasets naming as 20210825. The spe_dataset_20210818 remains the same.

In [28]:
bulk_p_df = pd.read_csv('data/bulk_dataset_20210818.csv', index_col=0)
print(bulk_p_df.shape)
c_df = pd.concat(
    [bulk_p_df[bulk_p_df.core != 'PS75-056-1'], bulk_df],
    axis=0, join='outer'
)
c_df.info()

(2522, 5)
<class 'pandas.core.frame.DataFrame'>
Index: 2522 entries, SO264-64-1_00115 to PS75-056-1_10155
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   TC%           2148 non-null   float64
 1   TOC%          2522 non-null   float64
 2   CaCO3%        2459 non-null   float64
 3   core          2522 non-null   object 
 4   mid_depth_mm  2522 non-null   int64  
dtypes: float64(3), int64(1), object(1)
memory usage: 118.2+ KB


In [29]:
c_df.to_csv('data/bulk_dataset_{}.csv'.format(date))

In [30]:
merge_p_df = pd.read_csv('data/spe+bulk_dataset_20210818.csv', index_col=0)
print(merge_p_df.shape)

c_df = pd.concat(
    [merge_p_df[merge_p_df.core != 'PS75-056-1'], merge_df],
    axis=0, join='outer'
)
c_df.info()

(2350, 2053)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2350 entries, 0 to 102
Columns: 2053 entries, 0 to mid_depth_mm
dtypes: object(2053)
memory usage: 36.8+ MB


In [32]:
c_df.to_csv('data/spe+bulk_dataset_{}.csv'.format(date))

In [31]:
print(date)

20210825
